pkg/synthetictests/resource_growth.go

```diff
@@ -0,0 +1,105 @@
+	return comparePostUpgradeResourceCountFromMetrics(testName, "secrets")
+
+}
+
+func testNoExcessiveConfigMapGrowthDuringUpgrade() []*junitapi.JUnitTestCase {
+	const testName = "[sig-trt] ConfigMap count should not have grown significantly during upgrade"
+	return comparePostUpgradeResourceCountFromMetrics(testName, "configmaps")
+
+}
+
+// comparePostUpgradeResourceCountFromMetrics tests that some counts for certain resources we're most interested
+// in potentially leaking do not increase substantially during upgrade.
+// This is in response to a bug discovered where operators were leaking Secrets and ultimately taking down clusters.
+// The counts have to be recorded before upgrade which is done in test/e2e/upgrade/upgrade.go, stored in a package
+// variable, and then read here in invariant. This is to work around the problems where our normal ginko tests are
+// all run in separate processes by themselves.
+//
+// Values for comparison are a guess at what would have caught the leak we saw. (growth of about 60% during upgrade)
+//
+// resource should be all lowercase and plural such as "secrets".
+func comparePostUpgradeResourceCountFromMetrics(testName, resource string) []*junitapi.JUnitTestCase {
+	oc := exutil.NewCLI("resource-growth-test")
+	ctx := context.Background()
+
+	preUpgradeCount := upgrade.PreUpgradeResourceCounts[resource]
+	e2e.Logf("found %d %s prior to upgrade", preUpgradeCount, resource)
+
+	cv, err := oc.AdminConfigClient().ConfigV1().ClusterVersions().Get(context.Background(), "version", metav1.GetOptions{})
+	if err != nil {
+		return []*junitapi.JUnitTestCase{
+			{
+				Name: testName,
+				FailureOutput: &junitapi.FailureOutput{
+					Output: "Error getting ClusterVersion: " + err.Error(),
+				},
+			},
+		}
+
+	}
+	upgradeCompletion := cv.Status.History[0].CompletionTime
+
+	// Use prometheus to get the resource count at the moment we recorded upgrade complete. We can't do this
+	// for the starting count as prometheus metrics seem to get wiped during the upgrade. We also don't want to
+	// just list the resources right now, as we don't know what other tests might have created since.
+	e2e.Logf("querying metrics at: %s", upgradeCompletion.Time.UTC().Format(time.RFC3339))
+	resourceCountPromQuery := fmt.Sprintf(`cluster:usage:resources:sum{resource="%s"}`, resource)
+	promResultsCompletion, err := helper.RunQueryAtTime(ctx, oc.NewPrometheusClient(ctx),
+		resourceCountPromQuery, upgradeCompletion.Time)
+	if err != nil {
+		return []*junitapi.JUnitTestCase{
+			{
+				Name: testName,
+				FailureOutput: &junitapi.FailureOutput{
+					Output: "Error getting resource count from Prometheus at upgrade completion time: " + err.Error(),
+				},
+			},
+		}
+
+	}
+	e2e.Logf("got %d metrics after upgrade", len(promResultsCompletion.Data.Result))
+	completedCount := float64(promResultsCompletion.Data.Result[0].Value)
+	e2e.Logf("found %f %s after upgrade", completedCount, resource)
```

Comment:
looking at the [e2e-gcp build-log.txt](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/27354/pull-ci-openshift-origin-master-e2e-gcp-builds/1559231208313851904/build-log.txt), I see:

```
Aug 15 19:42:20.153: INFO: found %!d(float64=0) Secrets prior to upgrade at 2022-08-15T19:42:20Z
Aug 15 19:42:20.190: INFO: upgrade ran from 2022-08-15T17:58:13Z -> 2022-08-15T18:25:46Z
Aug 15 19:42:20.190: INFO: querying metrics at: 2022-08-15T18:25:46Z
Aug 15 19:42:20.593: INFO: got 1 metrics after upgrade
Aug 15 19:42:20.593: INFO: found %!d(float64=924) Secrets after upgrade
```

I think it's showing 924 secrets (with some extra stuff in the string).  If 924 is the number, this can be considered a minor nit.

Reply:
Your timestamp there is from Aug 15, which was during development and those issues should now be resolved. Looks like that job was removed from this, and the failure still shows up in the latest result even though it's not run anymore because the last attempt failed. See https://prow.ci.openshift.org/pr-history?org=openshift&repo=origin&pr=27354

This code now should not run on any non-upgrade invocations of openshift-test as it's tied into the upgrade subcommand. Example: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/27354/pull-ci-openshift-origin-master-e2e-gcp/1562048845725044736/artifacts/e2e-gcp/openshift-e2e-test/build-log.txt does not show the "prior to upgrade" line at all, nor the test.

Reply:
ah, looks good:

```
Aug 23 14:27:12.374: INFO: got 1 metrics after upgrade
Aug 23 14:27:12.374: INFO: found 1124.000000 secrets after upgrade
Aug 23 14:27:12.374: INFO: found 586 configmaps prior to upgrade
Aug 23 14:27:12.410: INFO: querying metrics at: 2022-08-23T14:24:16Z
Aug 23 14:27:12.691: INFO: got 1 metrics after upgrade
Aug 23 14:27:12.691: INFO: found 672.000000 configmaps after upgrade
```

Reply:
Might be nicer if it was truncated, e.g. `%.2f`